cpp/include/cuspatial/experimental/detail/linestring_polygon_distance.cuh

```diff
@@ -0,0 +1,190 @@
+    auto num_segment_this_multilinestring =
+      multilinestrings.multilinestring_segment_count_begin()[geometry_id];
+    // The segment id from the multilinestring this thread is computing (local_id + global_offset)
+    auto multilinestring_segment_id =
+      local_idx % num_segment_this_multilinestring + multilinestrings_segment_offsets[geometry_id];
+    // The segment id from the multipolygon this thread is computing (local_id + global_offset)
+    auto multipolygon_segment_id =
+      local_idx / num_segment_this_multilinestring + multipolygons_segment_offsets[geometry_id];
+
+    auto [a, b] = multilinestrings.segment_begin()[multilinestring_segment_id];
+    auto [c, d] = multipolygons.segment_begin()[multipolygon_segment_id];
+
+    atomicMin(&distances[geometry_id], sqrt(squared_segment_distance(a, b, c, d)));
+  }
+};
+
+}  // namespace detail
+
+template <class MultiLinestringRange, class MultiPolygonRange, class OutputIt>
+OutputIt pairwise_linestring_polygon_distance(MultiLinestringRange multilinestrings,
+                                              MultiPolygonRange multipolygons,
+                                              OutputIt distances_first,
+                                              rmm::cuda_stream_view stream)
+{
+  using T       = typename MultiLinestringRange::element_t;
+  using index_t = iterator_value_type<typename MultiLinestringRange::geometry_it_t>;
+
+  CUSPATIAL_EXPECTS(multilinestrings.size() == multipolygons.size(),
+                    "Must have the same number of input rows.");
+
+  if (multilinestrings.size() == 0) return distances_first;
+
+  // Create a multipoint range from multilinestrings, computes intersection
+  auto multipoints           = multilinestrings.as_multipoint_range();
+  auto multipoint_intersects = point_polygon_intersects(multipoints, multipolygons, stream);
+
+  // Compute the "boundary" of threads. Threads are partitioned based on the number of linestrings
+  // times the number of polygons in a multipoint-multipolygon pair.
+  auto segment_count_product_it = thrust::make_transform_iterator(
+    thrust::make_zip_iterator(multilinestrings.multilinestring_segment_count_begin(),
+                              multipolygons.multipolygon_segment_count_begin()),
+    thrust::make_zip_function(thrust::multiplies<index_t>{}));
+
+  // Computes the "thread boundary" of each pair. This array partitions the thread range by
+  // geometries. E.g. threadIdx within [thread_bounds[i], thread_bounds[i+1]) computes distances of
+  // the ith pair.
+  auto thread_bounds = rmm::device_uvector<index_t>(multilinestrings.size() + 1, stream);
+  detail::zero_data_async(thread_bounds.begin(), thread_bounds.end(), stream);
+
+  thrust::inclusive_scan(rmm::exec_policy(stream),
+                         segment_count_product_it,
+                         segment_count_product_it + thread_bounds.size() - 1,
+                         thrust::next(thread_bounds.begin()));
```

Comment:
You've probably already thought of this but I see you wanting to have the thread partitioning code here and above refactored into a utility space for reuse in other algorithms.

Reply:
Yeah, this will come pretty handy when I start applying this parallel modality to other algorithms. I'm writing a refactor plan to st_distance and this will be factored out as part of that.